Do recursive update for commit when frame loading is complete.

ziransun opened the pull request:
#### 5364cc17ccbf86a58357b9d7a3d01655746fd170
<pre>
Do recursive update for commit when frame loading is complete.
<a href="https://bugs.webkit.org/show_bug.cgi?id=256976">https://bugs.webkit.org/show_bug.cgi?id=256976</a>

Reviewed by Chris Dumez.

In updateForCommit, we call recursiveUpdateForCommit to tell all
other frames to commit their provisional items and restore scroll
positions. This should only be called when the frame loading is complete.
This fixes an assertion failure for a couple of test cases with the combination
of cross document traversal and same document navigation.

* LayoutTests/TestExpectations:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::updateForCommit):
* Source/WebCore/loader/HistoryController.h:

Canonical link: <a href="https://commits.webkit.org/283904@main">https://commits.webkit.org/283904@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ba9e1a38f67b7efc6977cd6b4e27310a22586de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67606 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46985 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20238 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71655 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18743 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69724 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54783 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18547 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54147 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12532 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70673 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43121 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58465 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34608 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39794 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15870 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/block-container-block-start-child-with-border.html imported/w3c/web-platform-tests/css/css-grid/subgrid/subgrid-baseline-007.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17106 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61753 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16211 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73355 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11565 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15539 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61587 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58534 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61640 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15042 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9447 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3061 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42791 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43869 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45063 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43608 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/HistoryController.h:
```diff
@@ -114,6 +114,7 @@ class HistoryController final : public CanMakeCheckedPtr<HistoryController> {
     bool itemsAreClones(HistoryItem&, HistoryItem*) const;
     void updateBackForwardListClippedAtTarget(bool doClip);
     void updateCurrentItem();
+    bool isFrameLoadComplete() { return m_frameLoadComplete; }
```
`bool isFrameLoadComplete() const { return m_frameLoadComplete; }`

ziransun committed:
Do recursive update for commit when frame loading is complete.
https://bugs.webkit.org/show_bug.cgi?id=256976

Reviewed by Chris Dumez.

In updateForCommit, we call recursiveUpdateForCommit to tell all
other frames to commit their provisional items and restore scroll
positions. This should only be called when the frame loading is complete.
This fixes an assertion failure for a couple of test cases with the combination
of cross document traversal and same document navigation.

* LayoutTests/TestExpectations:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::updateForCommit):
* Source/WebCore/loader/HistoryController.h:

Canonical link: https://commits.webkit.org/283904@main
